mailet/standard/src/main/java/org/apache/james/transport/matchers/HasHeaderGlobalAndSpecific.java

```diff
@@ -0,0 +1,163 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+
+package org.apache.james.transport.matchers;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.StringTokenizer;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.mime4j.codec.DecodeMonitor;
+import org.apache.james.mime4j.codec.DecoderUtil;
+import org.apache.james.mime4j.util.MimeUtil;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMatcher;
+
+/**
+ * use:
+ *
+ * <pre>
+ * <code>&lt;mailet match="HasHeaderGlobalAndSpecific={&lt;header&gt;[=value]}+" class="..." /&gt;</code>
+ * </pre>
+ * <p/>
+ * <p>
+ * This matcher checks if the header is present in the message (global) and per recipient (specific). It complements the AddHeader mailet.
+ * </p>
+ */
+public class HasHeaderGlobalAndSpecific extends GenericMatcher {
+
+    private interface HeaderCondition {
+        Collection<MailAddress> isMatching(Mail mail) throws MessagingException;
+    }
+
+    private static class HeaderNameCondition implements HeaderCondition {
+        private final String headerName;
+
+        public HeaderNameCondition(String headerName) {
+            this.headerName = headerName;
+        }
+
+        @Override
+        public Collection<MailAddress> isMatching(Mail mail) throws MessagingException {
+            // Check global
+            String[] headerArray = mail.getMessage().getHeader(headerName);
+            if (headerArray != null && headerArray.length > 0) {
+                return mail.getRecipients();
+            }
+
+            // Check specific
+            Set<MailAddress> addressesWithHeader = new HashSet<>();
```

Comment:
We would tend to use Java 8 stream for doing this:

```
return mail.getPerRecipientSpecificHeaders().getHeadersByRecipient()
    .asMap()
    .entrySet()
    .stream()
    .filter(entry -> entry.getValue().getName().equals(headerName))
    .map(entry -> entry.getKey())
    .collect(Guavate.toImmutableSet());
```

Rationnals: functionnal like programming (without side effects + immutable datastructure) is easier to reason about and less bug-prone.